### PR TITLE
add blake2b/blake2s support for hmac

### LINKED ISSUE #3869: Cannot use BLAKE2b or BLAKE2s with HMAC
Python 3.6.2 on Ubuntu 17.10
Cryptography and dependencies installed via pip in virtualenv
cffi==1.10.0
cryptography==2.0.3
pip==9.0.1
setuptools==36.2.7

Steps to reproduce:
```
In [1]: from cryptography.hazmat.backends import default_backend
   ...: from cryptography.hazmat.primitives import hashes
   ...: from cryptography.hazmat.primitives.hmac import HMAC
   ...: 
   ...: backend = default_backend()
   ...: 

In [2]: hmac = HMAC(b'\x00'*32, hashes.SHA256(), backend)  # just fine

In [3]: hmac = HMAC(b'\x00'*32, hashes.BLAKE2s(digest_size=32), backend)
---------------------------------------------------------------------------
UnsupportedAlgorithm                      Traceback (most recent call last)
<ipython-input-3-61f273a52c45> in <module>()
----> 1 hmac = HMAC(b'\x00'*32, hashes.BLAKE2s(digest_size=32), backend)

~/.venvs/master/lib/python3.6/site-packages/cryptography/hazmat/primitives/hmac.py in __init__(self, key, algorithm, backend, ctx)
     30         self._key = key
     31         if ctx is None:
---> 32             self._ctx = self._backend.create_hmac_ctx(key, self.algorithm)
     33         else:
     34             self._ctx = ctx

~/.venvs/master/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/backend.py in create_hmac_ctx(self, key, algorithm)
    176 
    177     def create_hmac_ctx(self, key, algorithm):
--> 178         return _HMACContext(self, key, algorithm)
    179 
    180     def _build_openssl_digest_name(self, algorithm):

~/.venvs/master/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/hmac.py in __init__(self, backend, key, algorithm, ctx)
     32                     "{0} is not a supported hash on this backend.".format(
     33                         algorithm.name),
---> 34                     _Reasons.UNSUPPORTED_HASH
     35                 )
     36             res = self._backend._lib.HMAC_Init_ex(

UnsupportedAlgorithm: blake2s is not a supported hash on this backend.
```

### PR DESCRIPTION
**Pull request by reaperhulk**
This was a bug, but it turns out the noise protocol suggests using the HMAC construction with BLAKE2 (rather than BLAKE2's own keyed functionality) for a few reasons, so we should support it.

Fixes #3869

### CODE REVIEW

**Comment by reaperhulk** [comment]
The test vectors here were derived using python3.6 stdlib hmac + blake2. Python 3.6 uses OpenSSL for HMAC, but uses the C reference implementation of blake2, so we have at least some moderate assurance that OpenSSL is doing what we'd expect. Here's the code I used:

```python
import hashlib
import hmac

h = hmac.HMAC(key=b"0" * 64, digestmod=hashlib.blake2b)
h.update(b"test")
b2b = h.hexdigest()
h = hmac.HMAC(key=b"0" * 32, digestmod=hashlib.blake2s)
h.update(b"test")
b2s = h.hexdigest()
```

**Comment by alex** [comment]
That's not correct that it uses OpenSSL for HMAC. CPython has a pure-python HMAC on top of hashlib.

**Comment by reaperhulk** [comment]
Oh, even better! I'll edit my comment for posterity

File: tests/hazmat/primitives/test_hmac_vectors.py [hunk 1]
```diff
@@ -107,3 +109,29 @@ class TestHMACSHA512(object):
...
+        digest == binascii.unhexlify(
+            b"b5319122f8a24ba134a0c9851922448104e25be5d1b91265c0c68b22722f0f29"
+            b"87dba4aeaa69e6bed7edc44f48d6b1be493a3ce583f9c737c53d6bacc09e2f32"
+        )
+
+    def test_blake2s(self, backend):
+        h = hmac.HMAC(b"0" * 32, hashes.BLAKE2s(digest_size=32), backend)
+        h.update(b"test")
+        digest = h.finalize()
+        digest == binascii.unhexlify(
```
Comment (alex) [inline tests/hazmat/primitives/test_hmac_vectors.py]: You need an assert keyword here.

Comment (reaperhulk) [inline tests/hazmat/primitives/test_hmac_vectors.py]: lol

**Pull request merged**